binder/src/test/java/io/grpc/binder/RobolectricBinderSecurityTest.java

```diff
@@ -0,0 +1,414 @@
+import io.grpc.ClientCall;
+import io.grpc.ManagedChannel;
+import io.grpc.MethodDescriptor;
+import io.grpc.Server;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerMethodDefinition;
+import io.grpc.ServerServiceDefinition;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.protobuf.lite.ProtoLiteUtils;
+import io.grpc.stub.ClientCalls;
+import io.grpc.stub.ServerCalls;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.Robolectric;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.android.controller.ServiceController;
+
+@RunWith(RobolectricTestRunner.class)
+public final class RobolectricBinderSecurityTest {
+
+  private static final String SERVICE_NAME = "fake_service";
+  private static final String FULL_METHOD_NAME = "fake_service/fake_method";
+  private final Application context = ApplicationProvider.getApplicationContext();
+  private ServiceController<SomeService> controller;
+  private SomeService service;
+  private ManagedChannel channel;
+
+  @Before
+  public void setUp() {
+    controller = Robolectric.buildService(SomeService.class);
+    service = controller.create().get();
+
+    AndroidComponentAddress listenAddress = AndroidComponentAddress.forContext(service);
+    ScheduledExecutorService executor = service.getExecutor();
+    channel =
+        BinderChannelBuilder.forAddress(listenAddress, context)
+            .executor(executor)
+            .scheduledExecutorService(executor)
+            .offloadExecutor(executor)
+            .build();
+    idleLoopers();
+  }
+
+  @After
+  public void tearDown() {
+    channel.shutdownNow();
+    controller.destroy();
+  }
+
+  @Test
+  public void testAsyncServerSecurityPolicy_failed_returnsFailureStatus() throws Exception {
+    ListenableFuture<Status> status = makeCall();
+    service.setSecurityPolicyStatusWhenReady(Status.ALREADY_EXISTS);
+    idleLoopers();
+
+    assertThat(Futures.getDone(status).getCode()).isEqualTo(Status.Code.ALREADY_EXISTS);
+  }
+
+  @Test
+  public void testAsyncServerSecurityPolicy_allowed_returnsOkStatus() throws Exception {
+    ListenableFuture<Status> status = makeCall();
+    service.setSecurityPolicyStatusWhenReady(Status.OK);
+    idleLoopers();
+
+    assertThat(Futures.getDone(status).getCode()).isEqualTo(Status.Code.OK);
+  }
+
+  private void idleLoopers() {
```

Comment:
One more thing: is this actually correct? Suppose service.idleLooper() runs, then getMainLooper().idle() runs but the last thing it does is enqueue something to run on the test_handler_thread? Now idleLoopers() returns because it's idle but test_handler_thread is no longer idle. The two event loops could continue to pass work between themselves. And this kind of thing really happens. Like, when you make your first RPC, gRPC is going to call bindService() which enqueues the `onServiceConnected()` callback on the main thread. That callback then is going to enqueue the security policy check on the offload executor ...

Can you do everything on the single robolectric main thread?

Reply:
Updated. Now everything runs on the main thread.